Add feature to choose aggregates and percentiles in Histograms

gquinteros93 opened the pull request:
### Summary

This PR adds the optional feature to choose the aggregates and percentiles in the Histograms.

### Why

Because there are situations where we only wants to have a specific aggregate and/or percentiles, so the other data generated is extra data that we don't need

### Feature

This follows the code implemented on this [closed PR](https://github.com/dbader/node-datadog-metrics/pull/8). Thanks @cgiovanacci for starting this idea.

### Default

By default, if the user doesn't set the `option` value in the `histrogram` function, it will act as it is until now.

### Examples

1. Add a histogram with all the aggregates and only the `p75` percentile.

```
metrics.histogram('test.service_time', 0.248, 'myhost', { percentiles: [0.75] });
```

2. Add a histogram with the `max`, `min`, and `avg` aggregates and all the percentiles.

```
metrics.histogram('test.service_time', 0.248, 'myhost', { aggregates: ['max', 'min', 'avg'] });
```

3. Add a histogram with the `sum`, `count`, and `avg` aggregates and the `0.85`, `0.95` percentiles.

```
metrics.histogram('test.service_time', 0.248, 'myhost', { aggregates: ['max', 'min', 'avg'], percentiles: [0.85, 0.95] });
```

ErikBoesen commented on lib/metrics.js:
```diff
@@ -136,26 +140,39 @@ Histogram.prototype.addPoint = function(val, timestampInMillis) {
+Histogram.prototype.flush = function () {
+    let points = [];
+    if (this.aggregates.indexOf("min") !== -1) {
+      points.push(this.serializeMetric(this.min, "gauge", this.key + ".min"));
```
Nit: prefer single quotes for consistency with existing code

gquinteros93 replied:
Make sense 👍 . I changed all the `"` by `'`

ErikBoesen commented on lib/loggers.js:
```diff
@@ -80,8 +80,8 @@ BufferedMetricsLogger.prototype.increment = function(key, value, tags, timestamp
 
-BufferedMetricsLogger.prototype.histogram = function(key, value, tags, timestampInMillis) {
-    this.addPoint(Histogram, key, value, tags, timestampInMillis);
+BufferedMetricsLogger.prototype.histogram = function(key, value, tags, timestampInMillis, options) {
```
Could we add an optional `{}` value for `options`?

gquinteros93 replied:
Of course. Added 👍

ErikBoesen commented on lib/metrics.js:
```diff
@@ -136,26 +140,39 @@ Histogram.prototype.addPoint = function(val, timestampInMillis) {
-
-    const calcPercentile = function(p) {
-        const val = this.samples[Math.round(p * this.samples.length) - 1];
-        const suffix = '.' + Math.floor(p * 100) + 'percentile';
```
Four space tabs are used elsewhere, prefer adherence to that style

gquinteros93 replied:
Yes, that makes sense too. I updated the space tabs space used in that section.

ErikBoesen commented on test/metrics_tests.js:
```diff
@@ -196,4 +196,31 @@ describe('Histogram', function() {
+        const aggregates = ['avg'];
+        const percentiles = [0.85];
+        const h = new metrics.Histogram('hist', [], 'myhost',
+            { aggregates, percentiles});
```
Nit: spacing around `}`

gquinteros93 replied:
Added 👍

ErikBoesen commented:
Just some tiny stylistic changes—thanks for your work on this!

gquinteros93 commented:
> ErikBoesen

Hi @ErikBoesen, thanks for your reviewing the PR. 
I addressed all your comments.
Please let me know if there are any other improvements to apply.
